### ISSUE TITLE
Chunked FFTs

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#2620]
Admittedly this is a bit of a tricky topic, but it would be very helpful if we were able to perform FFTs over axes that have more than one chunk. This could help lighten the memory requirements in some cases and avoid potentially needing to rechunk the data.

**Comment by mrocklin** [comment]
I've always assumed that parallelizing individual FFT computations would be difficult to do well.  Do you know of efficient algorithms for this?

**Comment by martindurant** [comment]
One can do FFTs on the chunks and them combine them (Cooley-Tukey-like) with the "butterfly" operator (with twiddle factors, etc.), which is many-to-many; you would presumably need N=2**x where N is the number of chunks on the dimension and x is the number of shuffles/generations. This works particularly well when you can write the data back to the same allocated array space, but in our case probably means temporarily at least doubling memory use within a generation.
When I looked into this some time ago, 
I found the following two old papers that might be useful:
- C Calvin, 1996, Implementation of parallel FFT algorithms on distributed memory machines with a minimum overhead of communication.
- T H Cormen, 1997, Determining an Out-ofcore FFT Decomposition Strategy for Parallel Disks by Dynaic Programming

**Comment by roxyboy** [comment]
Would something like this help? `xrft` allows to take parallelized FFT along chunked axes. https://xrft.readthedocs.io/en/latest/chunk_example.html

**Comment by martindurant** [comment]
@roxyboy , no, this was about true, coherent transform, not averaging over the FFTs of chunks. The "statistical" FFT you would get by averaging may well still be useful, and indeed would be much easier to implement, but it should not be confused with a real FFT.

**Comment by jakirkham** [comment]
@grlee77, do you have any thoughts on how to approach this use case?

**Comment by grlee77** [comment]
@jakirkham, that does sound like a useful feature. Is this typically for multidimensional FFTs, batched 1D FFTs or for a single large 1D FFT?

I don't have experience with performing a single 1D FFT in a distributed fashion, but the references @martindurant sounds like it may be related to the approach described in the [multi-GPU 1D cuFFT docs](https://docs.nvidia.com/cuda/cufft/index.html#oned-mgpu-data-organization). I would have to think about it, but it seems plausible that Dask could do something similar, possibly with some restrictions on the chunk sizes.

@leofang added multi-GPU 1D FFTs to CuPy last year. It seems based on [this comment](https://github.com/cupy/cupy/pull/2644#issuecomment-558835428) that it is useful mainly when the data is too big to fit in memory.

**Comment by jakirkham** [comment]
Thanks Greg! Yeah I think this can come up with 1-D particularly if people are separating dimensions for an N-D case. This comes up in a few communities (biological imaging, medical imaging, satellite imaging, and larger scale modeling). The use cases are either wanting the result from the FFT itself or reusing it in something else like convolutions (or other operations reducible to a series of convolutions).

**Comment by grlee77** [comment]
Right, so the typical case you are talking about here are batched 1D FFTs where each individual one would fit in GPU memory. So you can do this operation now with Dask FFT, but performance is poor since you would have to repeatedly rechunk the data to have a single chunk along each axis in turn.

**Comment by jakirkham** [comment]
Yeah exactly rechunking is a bit painful atm. Avoiding rechunking would be ideal

**Comment by leofang** [comment]
I thought distributed FFT is a solved problem? Apart from Dask's (possibly) special chunking scheme, is there other substantial obstacle in adapting what's done in mpi4py-fft's approach?

https://mpi4py-fft.readthedocs.io/en/latest/

**Comment by grlee77** [comment]
Thanks @leofang, I had forgotten about mpi4py-fft (even though someone just mentioned a few days ago over at the pyFFTW repo!).

I just found this FluidFFT package dealing with distributed FFTs as well: https://fluidfft.readthedocs.io/en/latest/. It seems it also supports use of cuFFT for the FFTs, but I haven't looked at it closely yet.

**Comment by martindurant** [comment]
My original concept would be to start with a single 1D but very large array. Anything else should easily follow from that. Since communication will tend to be the bottleneck, there is less of a need to make any GPU-specific design choices. As noted, this has been done elsewhere, and should not be hard to express as a dask graph!

**Comment by grlee77** [comment]
It looks like @moble implemented an out-of-core FFT based on Cormen's approach (linked by @martindurant above) in https://github.com/moble/dask_fft.

It seems like the problem solved by mpi4py-fft and FluidFFT are a different case from the out-of-core approach, and instead deal with efficient domain decomposition of multi-dimensional transforms (e.g. slab or pencil decompositions in 3D). Unless I am mistaken, at any given time, one axis is contiguous in memory and standard FFTs are applied along that axis. Then a transpose is performed so that a different axis is contiguous in memory for the next dimension, etc.

**Comment by moble** [comment]
@grlee77 I didn't even remember I had that repo, and I can't say anything about whether the code in there works.  In particular, I've never been very good at using dask, so I kind of think I just gave up on that repo.

However, I have https://github.com/moble/out_of_core_fft, which was mostly written after that other repo.  I can't guarantee its current state (e.g., CI fails because of pytest updates), but I will say that it has worked very well for me and others in the past.

I should note that, because of the nature of this approach, there is potential for this to be *extremely* slow if memory-access patterns get weird.  I guess a dask chunk has similar issues to HDF5 chunks — which is what I was dealing with in that second repo.  But there are almost certainly use cases I didn't consider, so some thought should be put into generalizing it.

**Comment by leofang** [comment]
Thanks @grlee77 for the pointers to FluidFFT and the out-of-core implementation. My impression for mpi4py-fft aligns with your assessment. It seems straightforward to me to support cuFFT in mpi4py-fft because of the application of "standard FFT" (which includes numpy.fft, scipy.fft, pyfftw, etc), so I need to read more to understand if FluidFFT addresses other things not considered by mpi4py-fft. (I was interested in this cuFFT extension but got distracted by other things...)
